internal/clients/manager.go

```diff
@@ -154,102 +202,148 @@ func (c *AdminAPIClientsManager) SubscribeToGatewayClientsChanges() (<-chan stru
+// - readinessReconciliationTicker - triggered on every readinessReconciliationTicker tick.
+func (c *AdminAPIClientsManager) gatewayClientsReconciliationLoop() {
+	c.readinessReconciliationTicker.Reset(DefaultReadinessReconciliationInterval)
+	defer c.readinessReconciliationTicker.Stop()
+
+	close(c.running)
 	for {
 		select {
 		case <-c.ctx.Done():
 			c.logger.Infof("closing AdminAPIClientsManager: %s", c.ctx.Err())
-			c.discoveredAdminAPIsNotifyChan = nil
 			c.closeGatewayClientsSubscribers()
 			return
-
 		case discoveredAdminAPIs := <-c.discoveredAdminAPIsNotifyChan:
-			c.logger.Debug("received notification about Admin API addresses change")
-			if clientsChanged := c.adjustGatewayClients(discoveredAdminAPIs); clientsChanged {
-				// Notify subscribers that the clients list has been updated.
-				c.logger.Debug("notifying subscribers about gateway clients change")
-				c.notifyGatewayClientsSubscribers()
-			}
+			c.onDiscoveredAdminAPIsNotification(discoveredAdminAPIs)
+		case <-c.readinessReconciliationTicker.Channel():
+			c.onReadinessReconciliationTick()
 		}
 	}
 }
 
+// onDiscoveredAdminAPIsNotification is called when a new notification about Admin API addresses change is received.
+// It will adjust lists of gateway clients and notify subscribers about the change if readyGatewayClients list has
+// changed.
+func (c *AdminAPIClientsManager) onDiscoveredAdminAPIsNotification(discoveredAdminAPIs []adminapi.DiscoveredAdminAPI) {
+	c.logger.Debug("received notification about Admin API addresses change")
+
+	changed := func() bool {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		clientsChanged := c.adjustGatewayClients(discoveredAdminAPIs)
+		readinessChanged := c.reconcileGatewayClientsReadiness()
+		return clientsChanged || readinessChanged
+	}()
+
+	if changed {
+		c.notifyGatewayClientsSubscribers()
+	}
+}
+
+// onReadinessReconciliationTick is called on every readinessReconciliationTicker tick. It will reconcile readiness
+// of all gateway clients and notify subscribers about the change if readyGatewayClients list has changed.
+func (c *AdminAPIClientsManager) onReadinessReconciliationTick() {
+	c.logger.Debug("reconciling readiness of gateway clients")
+
+	changed := func() bool {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		return c.reconcileGatewayClientsReadiness()
+	}()
+
+	if changed {
+		c.notifyGatewayClientsSubscribers()
+	}
+}
+
 // adjustGatewayClients adjusts internally stored clients slice based on the provided
 // discovered Admin APIs slice. It consults BaseRootURLs of already stored clients with each
 // of the discovered Admin APIs and creates only those clients that we don't have.
 // It returns true if the gatewayClients slice has been changed, false otherwise.
 func (c *AdminAPIClientsManager) adjustGatewayClients(discoveredAdminAPIs []adminapi.DiscoveredAdminAPI) (changed bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	// Short circuit
+	// Short circuit.
```

Comment:
That's a fair point. I ended up with this kind of locking because initially I made `adjustGatewayClients` call `reconcileGatewayClientsReadiness` and I had to move the locking up. Now it doesn't make sense and I agree it's safer to keep locks on the level of every function that touches the mutex-protected data directly. 👍